### PR TITLE
feat(widgets): Add Typewriter widget

### LINKED ISSUE #478: feat(widgets): add Typewriter widget
> For AI agents (Claude Code, Copilot, Cursor): read [AGENTS.md](https://github.com/Karanjot786/TermUI/blob/main/AGENTS.md) and [packages/widgets/AGENTS.md](https://github.com/Karanjot786/TermUI/blob/main/packages/widgets/AGENTS.md) before writing code. This issue is a complete spec. Follow it exactly.

## Objective

Add a `Typewriter` widget to `@termuijs/widgets` that reveals static text character by character.

## Package scope

`packages/widgets`. All work is confined here. Import shared types from `@termuijs/core`.

## Files to create or modify

```
CREATE: packages/widgets/src/display/Typewriter.ts
CREATE: packages/widgets/src/display/Typewriter.test.ts
MODIFY: packages/widgets/src/index.ts   (add the Typewriter export)
```

## API contract

```typescript
import { Widget } from '../base/Widget.js';
import { type Style } from '@termuijs/core';

export interface TypewriterOptions {
    /** Characters revealed per tick. Default: 1 */
    speed?: number;
    /** Cursor glyph drawn at the reveal head. Default: caps-aware block. */
    cursor?: string;
}

export class Typewriter extends Widget {
    constructor(text: string, style?: Partial<Style>, opts?: TypewriterOptions);
    tick(): void;
    reset(): void;
    setText(text: string): void;
}
```

Export from `packages/widgets/src/index.ts`:

```typescript
export { Typewriter } from './display/Typewriter.js';
export type { TypewriterOptions } from './display/Typewriter.js';
```

## Acceptance criteria

- [ ] a new `Typewriter('hello')` reveals zero characters before the first `tick()`
- [ ] each `tick()` reveals `opts.speed` more characters until the full text is shown
- [ ] `tick()` past the end of the text is a no-op and does not exceed the full string
- [ ] a cursor glyph renders at the reveal head until the text is fully revealed
- [ ] `reset()` returns the revealed count to zero
- [ ] the cursor glyph uses `caps.unicode` with an ASCII fallback (for example `▋` falls back to `_`)
- [ ] `tick()`, `reset()`, and `setText()` call `markDirty()`
- [ ] `bun vitest run packages/widgets` passes
- [ ] `bun run typecheck` passes

## Test expectations

```
File: packages/widgets/src/display/Typewriter.test.ts
Run:  bun vitest run packages/widgets
```

Use the real `Screen` from `@termuijs/core`. Use `vi.spyOn` for caps, never mutate it.

Cases:
- reveals nothing before the first tick
- tick reveals more characters each call
- tick past the end stops at the full text
- reset returns to zero revealed
- ASCII fallback cursor renders when caps.unicode is false

```typescript
import { Screen, caps } from '@termuijs/core';
const screen = new Screen(20, 1);
const tw = new Typewriter('hello');
tw.updateRect({ x: 0, y: 0, width: 20, height: 1 });
tw.tick();
tw.tick();
tw.render(screen);
const row = screen.back[0].map(c => c.char).join('');
expect(row).toContain('he');
```

## Reference pattern to follow

Follow `packages/widgets/src/display/StreamingText.ts` for the reveal-counter pattern, the cursor glyph, and the `caps.unicode` fallback. Note `tick()` takes no arguments. Follow `packages/widgets/src/display/StreamingText.test.ts` for the test layout.

## Not included (out of scope)

- No internal timer; the host calls `tick()`.
- No word-wrap engine beyond simple truncation.
- No key-event handling.
- No JSX intrinsic registration.
- No docs-site changes.

## Do not touch

- Any package other than `packages/widgets`
- `packages/core` (import shared types, do not modify them)
- `bun.lock` (no dependency is needed)
- `.github/`

## Agent prompt (copy this into your AI agent)

```
Read AGENTS.md at the repo root and packages/widgets/AGENTS.md.
Read the reference file packages/widgets/src/display/StreamingText.ts and copy its structure.
Task: implement the Typewriter widget exactly to the API contract in this issue.
Create packages/widgets/src/display/Typewriter.ts and Typewriter.test.ts. Export Typewriter from packages/widgets/src/index.ts.
Plan first. List the files you will change and confirm they match the issue. Touch nothing outside packages/widgets. Do not edit bun.lock. Do not add dependencies.
Track a revealed counter and clamp it to the text length. Use caps.unicode for the cursor glyph with an ASCII fallback. Call markDirty() in tick, reset, and setText.
Tests use the real Screen and vi.spyOn for caps, never mutate caps.
When done run: bun vitest run packages/widgets && bun run typecheck. Both must pass.
```

## Before you open the PR

- [ ] I run `bun run build && bun vitest run && bun run typecheck` and all pass
- [ ] My change is confined to `packages/widgets`
- [ ] `bun.lock` has no unrelated changes
- [ ] I starred the repo

## GSSoC 2026

This is a beginner-level issue. Comment "I would like to work on this" to get assigned. You have 7 days to open a PR after assignment.

Read [CONTRIBUTING.md](./CONTRIBUTING.md) before starting.

### PR DESCRIPTION
**Pull request by bhanu-forge**
Description
Adds a Typewriter widget to @termuijs/widgets that reveals static text character by character. The host calls tick() at whatever cadence it likes — no internal timer is used. Includes reset() and setText() methods, a caps-aware cursor glyph, and 12 tests covering all acceptance criteria.
Related Issue

Closes #478
Which package(s)?
@termuijs/widgets
Type of Change

 ✨ Feature (type:feature)

Checklist

 ⭐ You starred the repo.
 Tests pass locally: bun vitest run
 Build passes: bun run build
 Typecheck passes: bun run typecheck
 You read CONTRIBUTING.md.
 Your PR title follows type: short description.
 Widget state mutators call markDirty() (tick, reset, setText all call it).
 No new any types without an inline comment explaining why.
 No unrelated refactors bundled into this PR.

GSSoC 2026 Participation
_

Screenshots / Recordings
N/A — terminal text widget, no visual diff needed.

Notes for the Reviewer
Three files touched, nothing outside packages/widgets:

src/display/Typewriter.ts — widget implementation
src/display/Typewriter.test.ts — 12 tests
src/index.ts — two export lines appended

## Summary by CodeRabbit

* **New Features**
  * Added a Typewriter display widget that reveals text incrementally with configurable speed and cursor, supports resetting and replacing text, respects available width (truncates to fit), and handles empty-width rendering.

* **Tests**
  * Added a comprehensive test suite covering reveal timing, cursor behavior (including unicode/ASCII fallback), reset/setText semantics, width truncation, and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Karanjot786/TermUI/pull/588?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/widgets/src/display/Typewriter.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  constructor(
+    text: string,
+    style: Partial<Style> = {},
+    opts: TypewriterOptions = {},
+  ) {
+    super(style);
+    this._text = text;
+    this._revealed = 0;
+    this._speed = opts.speed ?? 1;
+    // undefined means "auto from caps at render time"; a caller-supplied string
```
[the remark this answers is not in the document]
Comment (bhanu-forge) [reply packages/widgets/src/display/Typewriter.ts]: Fixed — speed is now clamped to a positive integer via Math.floor with a fallback to 1. / Fixed — truncation now uses stringWidth from @termuijs/core for display-width-aware slicing.

File: packages/widgets/src/display/Typewriter.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    // Visible text slice, truncated to content width.
+    const visibleText = this._text.slice(0, this._revealed);
+
+    // Append cursor when not yet fully revealed, but only if there is room.
+    const lineWithCursor = fullyRevealed
+      ? visibleText
+      : visibleText + cursorGlyph;
+
+    // Truncate to the available width so we never overflow the content rect.
+    const line = lineWithCursor.slice(0, width);
```
[the remark this answers is not in the document]
Comment (bhanu-forge) [reply packages/widgets/src/display/Typewriter.ts]: Fixed — speed is now clamped to a positive integer via Math.floor with a fallback to 1. / Fixed — truncation now uses stringWidth from @termuijs/core for display-width-aware slicing.
